sdcm/remote.py

```diff
@@ -897,12 +898,20 @@ def run(self, cmd, timeout=None, ignore_status=False,
         self.start_master_ssh()
 
         env = " ".join("=".join(pair) for pair in self.env.iteritems())
-        return self._run(cmd=cmd, timeout=timeout, verbose=verbose,
-                         ignore_status=ignore_status,
-                         connect_timeout=connect_timeout,
-                         env=env, options=options, args=args,
-                         log_file=log_file,
-                         watch_stdout_pattern=watch_stdout_pattern)
+        try:
+            return self._run(cmd=cmd, timeout=timeout, verbose=verbose,
+                             ignore_status=ignore_status,
+                             connect_timeout=connect_timeout,
+                             env=env, options=options, args=args,
+                             log_file=log_file,
+                             watch_stdout_pattern=watch_stdout_pattern)
+        except:
+            if retry <= 0:
+                raise
+            return self.run(cmd, timeout=timeout, ignore_status=ignore_status,
+                            connect_timeout=connect_timeout, options=options,
+                            verbose=verbose, args=args, log_file=log_file,
+                            watch_stdout_pattern=watch_stdout_pattern, retry=retry - 1)
```

Comment:
There is no need for recursion here. You can just do retry in loop, something like this:
`for try in range(retry):
    try:
        return self._run(...)
    except Exception:
        if try == retry - 1:
            raise
`

Reply:
I didn't use the for loop to avoid too nest indent.

Your example doesn't equal to the PR. In recursion retry, the code before self._run() will also be executed.

Reply:
What do you mean nest indent? If no exception happens, method will return immediately.
If you want any code before be executed, just move it inside the try clause.

Reply:
@amoskong, if we need the code before the _run we can move the try upper as Larisa mentioned.
Your code is probably working, but If we can avoid recursions I prefer to avoid them unless there is a strong case why using it (I didn't find one, but maybe you have).

Reply:
nesting statement isn't a big problem, let's use for loop as Larisa suggested.
https://codereview.stackexchange.com/questions/45220/how-do-i-avoid-deep-nesting-in-this-python-code

```
    def run(self, cmd, timeout=None, ignore_status=False,
        ....
        for try in range(retry):
            try:
                if args is None:
                    args = ()
                if verbose:
                    self.log.debug("Running '%s'", cmd)

                # Start a master SSH connection if necessary.
                self.start_master_ssh()

                env = " ".join("=".join(pair) for pair in self.env.iteritems())
                return self._run(cmd=cmd, timeout=timeout, verbose=verbose,
                                 ignore_status=ignore_status,
                                 connect_timeout=connect_timeout,
                                 env=env, options=options, args=args,
                                 log_file=log_file,
                                 watch_stdout_pattern=watch_stdout_pattern)
            except Exception:
                if try <= 0:
                    raise
```